src/java.desktop/share/classes/javax/swing/border/LineBorder.java

```diff
@@ -144,52 +143,18 @@ public LineBorder(Color color, int thickness, boolean roundedCorners)  {
      * @param height the height of the painted border
      */
     public void paintBorder(Component c, Graphics g, int x, int y, int width, int height) {
-        if ((this.thickness > 0) && (g instanceof Graphics2D)) {
-            Graphics2D g2d = (Graphics2D) g;
-
-            AffineTransform at = g2d.getTransform();
-
-            // if m01 or m10 is non-zero, then there is a rotation or shear
-            // or if no Scaling enabled,
-            // skip resetting the transform
-            boolean resetTransform = ((at.getShearX() == 0) && (at.getShearY() == 0)) &&
-                    ((at.getScaleX() > 1) || (at.getScaleY() > 1));
-
-            int xtranslation;
-            int ytranslation;
-            int w;
-            int h;
-            int offs;
-
-            if (resetTransform) {
-                /* Deactivate the HiDPI scaling transform,
-                 * so we can do paint operations in the device
-                 * pixel coordinate system instead of the logical coordinate system.
-                 */
-                g2d.setTransform(new AffineTransform());
-                double xx = at.getScaleX() * x + at.getTranslateX();
-                double yy = at.getScaleY() * y + at.getTranslateY();
-                xtranslation = clipRound(xx);
-                ytranslation = clipRound(yy);
-                w = clipRound(at.getScaleX() * width + xx) - xtranslation;
-                h = clipRound(at.getScaleY() * height + yy) - ytranslation;
-                offs = this.thickness * (int) at.getScaleX();
-            } else {
-                w = width;
-                h = height;
-                xtranslation = x;
-                ytranslation = y;
-                offs = this.thickness;
-            }
-
-            g2d.translate(xtranslation, ytranslation);
+        SwingUtilities3.paintBorder(c, g, x, y, width, height, this::paintUnscaledBorder);
+    }
 
-            Color oldColor = g2d.getColor();
-            g2d.setColor(this.lineColor);
+    private void paintUnscaledBorder(Component c, Graphics g, int x, int y,
+                                     int w, int h, double scale, int stroke) {
+        if (this.thickness > 0) {
```

Comment:
I believe this condition should remain the same:

```java
if ((this.thickness > 0) && (g instanceof Graphics2D)) {
```

After all, the border will not be painted if `g` isn't a `Graphics2D`, as such it makes sense to skip calculating the width and paths.

If you like, pattern matching for `instanceof` can still be used. I'm pretty sure @rajamah didn't use it to ease backporting.